src/Akeneo/Pim/Enrichment/Bundle/Storage/Sql/Completeness/SqlSaveProductCompletenesses.php

```
@@ -138,4 +131,48 @@ private function channelIdsIndexedByChannelCodes(): array
 
         return $result;
     }
+
+    /**
+     * To avoid to get several dead lock exceptions in a row, we sleep between the retry. It lets the database take a breath and finish the other concurrent transactions triggering the deadlock.
+     * There is a random sleep as well, to avoid to restart at the same time the other concurrent processes doing a retry as well.
+     */
+    private function executeWithRetry(callable $function): void
+    {
+        $retry = 0;
+        $isError = true;
+        while (true === $isError) {
+            try {
+                $this->connection->transactional($function);
+
+                $isError = false;
+            } catch (DeadlockException $e) {
+                $retry += 1;
+
+                usleep(300000 + rand(50000, $retry * 100000));
+
+                if (5 === $retry) {
+                    throw $e;
+                }
+            }
+        }
+    }
+
+    private function executeWithLockOnTable(callable $function): void
+    {
+        $value = $this->connection->executeQuery('SELECT @@autocommit')->fetch();
+        if (!isset($value['@@autocommit']) && ((int) $value['@@autocommit'] !== 1 || (int) $value['@@autocommit'] !== 0)) {
+            throw new \LogicException('Error when getting autocommit parameter from Mysql.');
+        }
+
+        $formerAutocommitValue = (int) $value['@@autocommit'];
+        try {
+            $this->connection->executeQuery('SET autocommit=0');
+            $this->connection->executeQuery('LOCK TABLES pim_catalog_completeness WRITE');
+            $function();
```

no catch?

---

we should at least log no?

---

no catch because in this case, that's the ultimate attempt. If we have an error, we should throw an exception.

---

maybe you should comment that